ARrehab/ARrehab/ViewController.swift

```diff
@@ -11,58 +11,247 @@ import RealityKit
+//Recieves updated scene info, can initiate actions based on scene-related events (i.e. anchor detection
+extension ViewController: ARSessionDelegate {
+        
     func session(_ session: ARSession, didAdd anchors: [ARAnchor]) {
-        if (hasMapped) {
-            return
+    }
+    
+    func session(_ session: ARSession, didUpdate anchors: [ARAnchor]) {
+        
+        if self.boardState == .notMapped {
+            for anc in anchors {
+                guard let planeAnc = anc as? ARPlaneAnchor else {break}
+                if self.boardState == .notMapped {
+                    if isValidSurface(plane: planeAnc) {
+                        self.initiateBoardLayout(surfaceAnchor: planeAnc)
+                    }
+                }
+            }
         }
-        var anc: ARAnchor?
-        anchors.forEach {anchor in
-            guard let planeAnchor = anchor as? ARPlaneAnchor else {return}
-            if (planeAnchor.alignment == .horizontal) { // TODO: change to classification == .floor, Get the planeAnchor to make sure that the plane is large enough
-                anc = planeAnchor
-                self.hasMapped = true
+            
+        else if self.boardState == .mapped {
+            for anc in anchors {
+                if anc == self.tileGrid?.surfaceAnchor {
+                    let planeAnc = anc as! ARPlaneAnchor
+                    self.tileGrid?.updateBoard(updatedAnc: planeAnc)
+                }
             }
         }
-        if (hasMapped) {
-            let ancEntity = AnchorEntity(anchor: anc!)
-            for x in -1 ... 1 {
-                for z in -1 ... 1 {
-                    let tile: Tile = Tile(name: String(format: "Tile (%d,%d)", x, z), x: Float(x)/2.0, z: Float(z)/2.0)
-                    ancEntity.addChild(tile)
+        
+    }
+    
+    func initiateBoardLayout(surfaceAnchor: ARPlaneAnchor) {
+        guard self.boardState == .notMapped else {return}
+        self.boardState = .mapping
+        
+        self.tileGrid = TileGrid(surfaceAnchor: surfaceAnchor)
+        self.arView.scene.addAnchor(self.tileGrid!.gridEntity)
+        self.boardState = .mapped
+        
+        self.addPbButton()
+        self.startBoardPlacement()
+    }
+    
+    func startBoardPlacement() {
+        
+        self.arView.scene.addAnchor(playerEntity)
+        self.playerEntity.addCollision()
+        
+        Timer.scheduledTimer(withTimeInterval: TimeInterval(exactly: 1.0)!, repeats: true) {timer in
+            
+            if self.boardState == .placed {
+                timer.invalidate()
+            }
+                
+            else if (self.playerEntity.onTile != nil) {
+                self.tileGrid?.updateBoardOutline(centerTile: self.playerEntity.onTile)
+            }
+            
+            
+        }
+
+        /*Timer.scheduledTimer(withTimeInterval: TimeInterval(exactly: 1.0)!, repeats: true, block: {timer in
```

Comment:
Is this the directional code?

Reply:
Yah, kind of a mess. Maybe you can make more sense of how to extract and use direction. From what I could find online, the transform matrix encodes the translation in the fourth column (this I can confirm -- checked it against given cameraTransform property of arView) and the direction in the third column (I can also confirm this, as when I was testing out the first method of getting direction (var direction) it was correctly tracing the camera. However, due to time constraints I dropped that path (one annoying issue was when projecting a ray from the camera, it would hit a whole list of tiles). Then there's also the rotation quaternion that I guess is a common representation of direction in 3D modeling... I tried understanding how to properly extract direction from that but dropped it as well. I can delete all this, but I think it would be good for us to eventually figure out direction.

Reply:
Yes please keep it. It will be good to work out of. This would be a nice element of our Summer project.